sound/soc/sof/debug.c

```diff
@@ -336,6 +336,103 @@ static int sof_debug_ipc_flood_test(struct snd_sof_dev *sdev,
 }
 #endif
 
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_IPC_MSG_INJECTOR)
+static ssize_t msg_inject_read(struct file *file, char __user *buffer,
+			       size_t count, loff_t *ppos)
+{
+	struct snd_sof_dfsentry *dfse = file->private_data;
+	struct sof_ipc_reply *rhdr = dfse->msg_inject_rx;
+
+	if (!rhdr->hdr.size || !count || *ppos)
+		return 0;
+
+	if (count > rhdr->hdr.size)
+		count = rhdr->hdr.size;
+
+	if (copy_to_user(buffer, dfse->msg_inject_rx, count))
+		return -EFAULT;
+
+	*ppos += count;
+	return count;
+}
+
+static ssize_t msg_inject_write(struct file *file, const char __user *buffer,
+				size_t count, loff_t *ppos)
+{
+	struct snd_sof_dfsentry *dfse = file->private_data;
+	struct snd_sof_dev *sdev = dfse->sdev;
+	struct sof_ipc_cmd_hdr *hdr = dfse->msg_inject_tx;
+	size_t size;
+	int ret, err;
+
+	if (!count || *ppos || count > SOF_IPC_MSG_MAX_SIZE)
+		return 0;
+
+	size = simple_write_to_buffer(dfse->msg_inject_tx, count, ppos, buffer,
+				      count);
+	memset(dfse->msg_inject_tx + count, 0, SOF_IPC_MSG_MAX_SIZE - count);
```

Comment:
having verified, that `size == count`, this shouldn't be needed?

Reply:
The message itself still can say that `hdr.size > count`, I want to make sure that in this case we _know_ that the bytes after count is always 0.

Reply:
why do you need those bytes as 0? Any specific reason for that?

Reply:
@lyakh, I don't want them to be random. It would be impossible to reproduce and fix a failure if we send random and not recorded values to the firmware.
If we want to test random values in messages then the message should be crafted that way.

Reply:
sure, no, we shouldn't be sending random data. But are those bytes sent in the IPC? Aren't only `count` bytes sent in lines 388-389 below?

Reply:
@lyakh, aaah, if you put it that way... You are right, the memset is pointless unless we use `SOF_IPC_MSG_MAX_SIZE` in place of count.
We can either drop the memset or make the core to copy the full msg_buffer to mailbox unconditionally.
What do you think?

Reply:
as long as there's no functional requirement or argument in favour of copying the whole buffer, I'd limit to only the required size.

Reply:
OK, I'll drop the memset for now. It can be added back if we saw it is needed.

Do you have other concerns before I push an update?